src/backend/tls.rs

```diff
@@ -0,0 +1,247 @@
+        }
+    }
+}
+
+pub struct TlsGenerate {
+    ca_cert: X509,
+    ca_key: PKey<Private>,
+}
+
+impl TlsGenerate {
+    pub fn new_self_signed() -> Self {
+        let subject_name = X509NameBuilder::new()
+            .and_then(|mut name| {
+                name.append_entry_by_nid(Nid::COMMONNAME, "secret-operator self-signed")?;
+                Ok(name)
+            })
+            .unwrap()
+            .build();
+        let now = OffsetDateTime::now_utc();
+        let not_before = now - Duration::minutes(5);
+        let not_after = now + Duration::days(2 * 365);
+        let conf = Conf::new(ConfMethod::default()).unwrap();
+        let ca_key = PKey::try_from(Rsa::generate(2048).unwrap()).unwrap();
+        let ca_cert = X509Builder::new()
+            .and_then(|mut x509| {
+                x509.set_subject_name(&subject_name)?;
+                x509.set_issuer_name(&subject_name)?;
+                x509.set_not_before(Asn1Time::from_unix(not_before.unix_timestamp())?.as_ref())?;
+                x509.set_not_after(Asn1Time::from_unix(not_after.unix_timestamp())?.as_ref())?;
+                x509.set_pubkey(&ca_key)?;
+                let mut serial = BigNum::new()?;
+                serial.rand(64, MsbOption::MAYBE_ZERO, false)?;
+                x509.set_serial_number(Asn1Integer::from_bn(&serial)?.as_ref())?;
+                x509.set_version(
+                    3 - 1, // zero-indexed
+                )?;
+                let ctx = x509.x509v3_context(None, Some(&conf));
+                let exts = [
+                    BasicConstraints::new().critical().ca().build()?,
+                    SubjectKeyIdentifier::new().build(&ctx)?,
+                    AuthorityKeyIdentifier::new()
+                        .issuer(false)
+                        .keyid(false)
+                        .build(&ctx)?,
+                    KeyUsage::new()
+                        .critical()
+                        .digital_signature()
+                        .key_cert_sign()
+                        .crl_sign()
+                        .build()?,
+                ];
+                for ext in exts {
+                    x509.append_extension(ext)?;
+                }
+                x509.sign(&ca_key, MessageDigest::sha256())?;
+                Ok(x509)
+            })
+            .unwrap()
+            .build();
+        Self { ca_key, ca_cert }
+    }
+
+    pub async fn get_or_create_k8s_certificate(
+        client: &stackable_operator::client::Client,
+    ) -> Self {
+        let k8s_secret_name = "secret-provisioner-ca";
+        let k8s_ns = "default";
+        let existing_secret = client.get::<Secret>(k8s_secret_name, Some(k8s_ns)).await;
+        match existing_secret {
+            Ok(ca_secret) => {
+                let ca_data = ca_secret.data.unwrap_or_default();
+                Self {
+                    ca_key: PKey::private_key_from_pem(&ca_data.get("ca.key").unwrap().0).unwrap(),
```

Comment:
Not sure what would be the correct behavior here, we probably don't want to throw away anything in the secret and regenerate a ca just because there was a maybe spurious parsing failure.
But at least returning an error would be better?

Reply:
Hence why we `create()` the `Secret` rather than `apply()`ing it. If it already exists then that should return with a conflict failure, which we present to the user. We might want to add a logging statement or something here though, to help understand why the creation path is being invoked in the first place?

Reply:
My comment was mostly referring to the unwraps that might fail if for example "ca.key" was mistyped in the Secret.

But these have been fixed in #21 so you can disregard this comment :)